core/modules/replica/IndexJob.h

```diff
@@ -0,0 +1,292 @@
+    typedef std::shared_ptr<IndexJob> Ptr;
+
+    /// The function type for notifications on the completion of the request
+    typedef std::function<void(Ptr)> CallbackType;
+
+    /// @return default options object for this type of a request
+    static Job::Options const& defaultOptions();
+
+    /// @return the unique name distinguishing this class from other types of jobs
+    static std::string typeName();
+
+    /// Possible destinations where the harvested data would go
+    enum Destination {
+        DISCARD,    // do nothing with the data
+        FILE,       // write all data into a file
+        FOLDER,     // write each chunk's data as a separate file at a folder
+        TABLE       // write into the specified or standard "secondary index" table
+    };
+
+    /// @return the string representation for a value of the Destination option
+    static std::string toString(Destination destination);
+
+    /// @return a value of the enumerator Destination parsed from the input string
+    /// @throw invalid_argument if the input value doen't match any known option of
+    ///        the enumerator type
+    static Destination fromString(std::string const& str);
+
+    /**
+     * Static factory method is needed to prevent issue with the lifespan
+     * and memory management of instances created otherwise (as values or via
+     * low-level pointers).
+     *
+     * @param database
+     *   the name of a database for which the "secondary index" is built
+     *
+     * @param hasTransactions
+     *   if 'true' then the database's "director" tables are expected to be partitioned,
+     *   and the job will extract data (including column "qserv_trans_id") from
+     *   a specific MySQL partition.
+     *
+     * @param transactionId
+     *   an identifier of a super-transaction which would limit a scope of the data
+     *   extraction requests. NOte this request will be considered only if
+     *   'hasTransactions=true'.
+     *
+     * @param allWorkers
+     *   engage all known workers regardless of their status. If the flag
+     *   is set to 'false' then only 'ENABLED' workers which are not in
+     *   the 'READ-ONLY' state will be involved into the operation.
+     *
+     * @param destination
+     *   possible destinations for the harvested data
+     * 
+     * @param destinationPath
+     *   depending on a value of the previous parameter 'destination', a value
+     *   of this parameter could be either either the name of a file, the name
+     *   of an existing folder, or the name of a table. For the FILE destination
+     *   the empty destination path will trigger dumping the data onto the Standard
+     *   Output Stream. For the FOLDER option the current working directory will
+     *   be assumed. And for the TABLE option the empty value would imply
+     *   the standard "secondary index" table of the database. A non-empty value
+     *   for the table would imply the name of a specific (non-standard) table.
```

Comment:
I think this would be a bit of an overkill for the class-level API ... at this level of the code. Packaging into strings and then unpacking would go against the type safety enforcements of the (C++) language.